src/Z3/Z3Node.class.st

```diff
@@ -247,6 +276,57 @@ Z3Node >> isNode [
 	^true
 ]
 
+{ #category : #'fx expr classification' }
+Z3Node >> isPAll [
+	"Answer whether Fixpoint would consider this node a PAll."
+	^false
+]
+
+{ #category : #'fx expr classification' }
+Z3Node >> isPAnd [
+	"Answer whether Fixpoint would consider this node a PAnd."
+	^false
+]
+
+{ #category : #'fx expr classification' }
+Z3Node >> isPAtom [
+	"Answer whether Fixpoint would consider this node a PAtom.
+	 A PAtom is (Brel e₁ ₂), where Brel = Eq | Ne | Gt | Ge | Lt | Le | Ueq | Une.
+	 In Z3 we don't have some of that, let's postpone deciding what to do there;
+	 it should never happen if callers check for isPNot first."
```

Comment:
I do understand the comment:

> In Z3 we don't have some of that, let's postpone deciding what to do there;

To where? 

And, if 

> it should never happen

shouldn't we just do `self error:` or maybe `self shouldBeImplemented` so we get an error when we hit this "should never happen" case?

Reply:
Alas, this is a very unclear area.  Obviously from Z3's perspective, `Ne` is just meaningless (there is simply no such relation).  But if the callers agree to always check for `isPNot` before checking for `isPAtom`, then this becomes a moot point.

The real problem is the horrible chain of `if-then-elseif-then-…` in https://github.com/shingarov/MachineArithmetic/blob/reflect5/src/PLE/Z3Node.extension.st#L11 and we can't do any better here, say for example dispatch on subclass, because the Smalltalk subclasses are structured by sort (`Int` vs `BitVector` vs `Bool` etc) and not by `AST_KIND`.  Even if there were some nice language-level concept for this, it would have to be at "refined" level (e.g. `AST_KIND=App` such that `functorName='='`).  If you look at the upstream counterpart: https://github.com/ucsd-progsys/liquid-fixpoint/blob/develop/src/Language/Fixpoint/Solver/PLE.hs#L570 and https://github.com/shingarov/liquid-fixpoint/blob/develop/src/Language/Fixpoint/Solver/PLE.hs#L494 you will see they are dispatched per data constructor but textually they are still all grouped together, whereas if we had Smalltalk-style one-method-per-class dispatch it would be scattered in many places.

> shouldn't we just do self error: or maybe self shouldBeImplemented

Well that's the thing, there isn't a place in the Smalltalk control where this "can't happen" is located, because there is no way in Z3 to construct an `Ne` node.

So I think from that perspective there is no much we can do differently.

Reply:
I hope you don't mind if I merge this now.  We've got bigger fish to fry right now.